test/e2e/storage/testsuites/base.go

```diff
@@ -95,16 +95,50 @@ var CSISuites = append(BaseSuites,
 	InitSnapshottableStressTestSuite,
 )
 
-// TestSuite represents an interface for a set of tests which works with TestDriver
-type TestSuite interface {
-	// GetTestSuiteInfo returns the TestSuiteInfo for this TestSuite
-	GetTestSuiteInfo() TestSuiteInfo
-	// DefineTests defines tests of the testpattern for the driver.
+// TestSuiteInterface represents an interface for a set of tests which works with TestDriver.
+// Each testsuite should implement this interface and use a TestSuiteHandler to define its Tests.
+// The function defined inside should be called only by TestSuiteHandler.
+// All the functions are designed to not be exported so that caller has to use TestSuiteHandler to
+// actually define tests.
+type TestSuiteInterface interface {
+	getTestSuiteInfo() TestSuiteInfo
+	// defineTests defines tests of the testpattern for the driver.
 	// Called inside a Ginkgo context that reflects the current driver and test pattern,
 	// so the test suite can define tests directly with ginkgo.It.
-	DefineTests(TestDriver, testpatterns.TestPattern)
-	// SkipRedundantSuite will skip the test suite based on the given TestPattern and TestDriver
-	SkipRedundantSuite(TestDriver, testpatterns.TestPattern)
+	defineTests(TestDriver, testpatterns.TestPattern)
+	// skipUnsupportedTests will skip the test suite based on the given TestPattern, TestDriver
+	// Testsuite should check if the given pattern and driver works for the "whole testsuite"
+	// Testcase specific check should happen inside defineTests
+	skipUnsupportedTests(TestDriver, testpatterns.TestPattern)
+}
+
+// TestSuiteHandler is the actual data struct we use to define tests
+// All the test registration should happen from the handler
+type TestSuiteHandler struct {
+	testSuite TestSuiteInterface
+}
+
+// GetTestSuiteInfo returns the TestSuiteInfo for a given testsuite wrapper
+func (h *TestSuiteHandler) GetTestSuiteInfo() TestSuiteInfo {
+	return h.testSuite.getTestSuiteInfo()
+}
+
+// RegisterTests register the driver + pattern combination to the inside TestSuite
+// This function actually register tests inside testsuite
+func (h *TestSuiteHandler) RegisterTests(driver TestDriver, pattern testpatterns.TestPattern) {
```

Comment:
If we don't have a `TestSuiteHandler`, it seems like this function could be refactored to remove the receiver, to take a `testSuite` as an argument, and still perform the same work of calling `testSuite.skipUnsupportedTests()`.

Reply:
And it would still be called in `DefineTestSuites()` and pass in a `TestSuite` interface as an argument.

Reply:
Yeah you are right. I am over complify this. Originally I was planning to have the testsuiteinfo also separate from the testsuite, so keep the testsuiteinterface as easy as possible. That's why I have this Handler class. Then I find out it is not necessary. So I remove the testsuiteinfo.

Good suggestion! I will do a refactor again

Reply:
Let me know if there's issues with that design, I may have missed something.